### ISSUE TITLE
Cache loading from the disk after server restart

### ISSUE THREAD

**Issue by ironsmile** [ironsmile/nedomi#9]
At the moment nedomi loses all of its cache on restart. I think is the job of the storage module to load its cache at server start.

An interface should be created "LoadCacher" for the function "LoadCache". If a storage implements it, it must be called on server startup.

Obviously we will have to add such a function to our current disk storage implementation.

**Comment by MStoykov** [comment]
The current disk implementation does not actually have any knowledge of what it has.
It asks the cache module if a piece is cached - if it is it's read from disk (failure is bad), if not it gets it through the upstream module. It only knows if something is currently being downloaded from the upstream or is serving something.
So it's the cache module that needs to save it's state and then load it not the storage module :) Atleast with the current implementations

**Comment by ironsmile** [comment]
This is all true.

But my idea is to abandon the cache module state after restart. We just have to make sure the cache knows what files can be found in the storage at startup. After all, files may have been removed or otherwise tampered with while the server has been stopped. I have one more reason to leave this work to for storage. Saving the cache state will require some storing routine to work continuously during runtime and we will have to implement a state loading as well. These two things will probably be different for every cache implementation. My feeling is that `LoadCacher` will be easier to do and maintain.

Writing all of this gave me a different idea. We may do it like this: never load the cache from storage. On every request for a file (or a part of it) we will check if it has been cached before. For a disk based storage this means a `stat` call. If it is and it is not stale yet - use it. The downside of this approach is that we might end up eating more space than configured in the cache zone. This is because some cached files will probably never be hit again.

**Comment by MStoykov** [comment]
I'm strictly against the "different idea". 

Given the way the current disk storage writes it's files, if the same configuration again it should be possible to both detect it and restore the storage.

**Comment by MStoykov** [comment]
Hm I think it will be better (I can more easily implement it) If just when storage.disk is created it checks whether it can load something and if it can it just calls cache.PromoteObject for each index that can be found. 

This will require to write the headers to the filesystem as I don't know how big files are (maybe the server was stopped while piece 15 of 30 was downloading). 
It would also require the cache module to be ready to promoteObject while the storage is still being initialized.
All of this gets more complicated if any part of the configuration is changed:
- on partSize change I suppose we can fix some parts. I don't know if it will be worth it. 
- making the amount of object smaller than before and then trying to load an already "full" cache is interesting situation. 
- given that the upstream could be changed, It will probably be good idea if all of that (and more) is written at the storage path for after restart comparison .

**Comment by ironsmile** [comment]
Yes, calling `PromoteObject` should be enough. In the normal situation all objects will be within their cache size limit and they will be stored in the cached.

And now for the less than normal cases. We may store the cache parameters in a "meta" file. The parameters will include the partSize, cacheSize and cacheKey.

If the partSize and cacheKey are different we might as well discard the whole thing. Maybe in background job.

Probably the most prominent difference will in the cacheSize. If it more than the one in the meta file - great! Just update the file and continue. If it is less we should remove some of the files until we shrink it back to the desired size. This last part is the difficult part. What file should be remove? My guess is random should be enough.

Hopefully this will make restarts without any changes easy and restarts with changes predictable.

**Comment by MStoykov** [comment]
Currently the storage have no way of knowing the CacheKey at startup, but even then it will not be enough. 

If this is part of vhost settings:
```json
           {
                "name": "127.0.0.1",
                "upstream_address": "http://inv.bg",
                "cache_zone": "zone2",
                "cache_key": "2.1"
            },
            {
                "name": "cached.doycho.com",
                "upstream_address": "http://localhost",
                "cache_zone": "zone2",
                "cache_key": "1.2"
            },
```
it's possible to change just the `cache_key` or change the `upstream_address` of one of them to one that is not *logically* the same host.

**Comment by na--** [comment]
I think that the upstream address should be used as a cache key directly or that it should automatically be used as a part of it.

**Comment by ironsmile** [comment]
A change to the `cache_key` is... problematic. But I don't see the change of the `upstream_address` as a notable event. You should be able to change your upstream address (say, because you need it to run on different port) and still retain your cache. This is something the writer of the config should assume responsibility for.

Actually, we should apply the same logic to the cache key issue. Sure, this zone has new cache key after restart. That says to me that the writer of the config wanted new cache to begin with. The files with the old cache key should be loaded as normal and they will be gradually replaced by new files given enough time.

**Comment by MStoykov** [comment]
I agree with the idea that the config writer should be responsible. And the question was mostly if the two keys are switched (in the above example) then wrong data will be delivered, and this can't be easily fixed at the current time and I can't come up with good solution.

Changing the upstream_address and keeping the cache seems like something that could happen (if the address is changed for example :tongue: ), so it should probably not be taken into account.

**Comment by ironsmile** [comment]
Closed with d688768.
